airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/conditional_paginator.py

```diff
@@ -2,40 +2,179 @@
+        url_base: Union[InterpolatedString, str],
+        decoder: Decoder = None,
+    ):
+        """
+        :param request_options_provider: additional request options to set
+        :param page_token_option: request option to set the the next_page_token
+        :param pagination_strategy: Strategy defining how to get the next page token
+        :param config: connection config
+        :param url_base: endpoint's base url
+        :param decoder: decoder to decode the response
+        """
+        self._request_options_provider = request_options_provider
         self._config = config
+        self._page_token_option = page_token_option
+        self._pagination_strategy = pagination_strategy
+        self._token = None
+        if isinstance(url_base, str):
+            url_base = InterpolatedString(url_base)
+        self._url_base = url_base
+        self._decoder = decoder or JsonDecoder()
 
     def next_page_token(self, response: requests.Response, last_records: List[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
+        """
+        Returns a mapping {"next_page_token": <token>}
+        This is required to tell HttpStream there is another page to request.
+
+        :param response: to process
+        :param last_records: records extracted from the response
+        :return: mapping {"next_page_token": token}
+        """
+        if self.stop_condition(response, last_records):
+            return None
+        else:
+            self._token = self._pagination_strategy.next_page_token(response, last_records)
+            if self._token:
+                return {"next_page_token": self._token}
+            else:
+                return None
+
+    @abstractmethod
+    def stop_condition(self, response: requests.Response, last_records: List[Mapping[str, Any]]) -> bool:
+        """
+        Predicate evaluating to True when there are no more pages to request
+        :param response:
+        :param last_records:
+        :return: boolean indicating whether we should stop paginating
+        """
+
+    def path(self):
+        if self._token and self._page_token_option.option_type == RequestOptionType.path:
+            # Replace url base to only return the path
+            return str(self._token).replace(self._url_base.eval(self._config), "")
+        else:
+            return None
+
+    def request_params(self) -> Mapping[str, Any]:
+        return {
+            **self._get_request_options(RequestOptionType.request_parameter),
+            **self._request_options_provider.request_params(stream_state=None, stream_slice=None, next_page_token=None),
+        }
+
+    def request_headers(self) -> Mapping[str, str]:
+        return {
+            **self._get_request_options(RequestOptionType.header),
+            **self._request_options_provider.request_headers(stream_state=None, stream_slice=None, next_page_token=None),
+        }
+
+    def request_body_data(self) -> Union[Mapping[str, Any], str]:
```

Comment:
as per in person convo, we'll get rid of the union and just return a mapping for simplicity

Reply:
agreed that's a better approach